src/libraries/System.Net.Http/src/System/Net/Http/Headers/HeaderDescriptor.cs

```diff
@@ -15,32 +16,91 @@ namespace System.Net.Http.Headers
     // Use HeaderDescriptor.TryGet to resolve an arbitrary header name to a HeaderDescriptor.
     internal readonly struct HeaderDescriptor : IEquatable<HeaderDescriptor>
     {
-        private readonly string _headerName;
-        private readonly KnownHeader? _knownHeader;
+        /// <summary>
+        /// Either a <see cref="KnownHeader"/> or <see cref="string"/>.
+        /// Marked as nullable since a default (uninitialized) instance of this struct is also used in practice
+        /// to indicate the end of the header collection.
+        /// </summary>
+        private readonly object? _descriptor;
 
         public HeaderDescriptor(KnownHeader knownHeader)
         {
-            _knownHeader = knownHeader;
-            _headerName = knownHeader.Name;
+            _descriptor = knownHeader;
         }
 
         // This should not be used directly; use static TryGet below
         internal HeaderDescriptor(string headerName)
         {
-            _headerName = headerName;
-            _knownHeader = null;
+            _descriptor = headerName;
         }
 
-        public string Name => _headerName;
-        public HttpHeaderParser? Parser => _knownHeader?.Parser;
-        public HttpHeaderType HeaderType => _knownHeader == null ? HttpHeaderType.Custom : _knownHeader.HeaderType;
-        public KnownHeader? KnownHeader => _knownHeader;
+        public bool HasValue => _descriptor is not null;
+
+        public string Name
+        {
+            get
+            {
+                Debug.Assert(_descriptor is not null);
+                object descriptor = _descriptor;
+                return descriptor is KnownHeader knownHeader ?
+                    knownHeader.Name :
+                    Unsafe.As<string>(descriptor);
+            }
+        }
+
+        public bool IsKnownHeader([NotNullWhen(true)] out KnownHeader? knownHeader, [NotNullWhen(false)] out string? headerName)
```

Comment:
Instead of using `Unsafe.SkipInit` here, consider passing as `ref` and putting `Unsafe.SkipInit` in the calling method. This will improve maintainability in the case a caller assumes the `out` to be defaulted later.

Reply:
I will incorporate this into the `HeaderDescriptor` change in a separate PR - https://github.com/dotnet/runtime/pull/62981#issuecomment-1006897535